src/main/java/org/spongepowered/api/resource/ResourceProvider.java

```diff
@@ -38,4 +39,18 @@
      */
     Optional<Resource> getResource(ResourcePath path);
 
+    /**
+     * Returns all of the resources which exist.
+     * @return
+     */
+    Collection<Resource> getResources();
+
+    /**
+     * Walks a resource tree starting at the given path.
+     *
+     * @param path The path to start at
+     * @return A stream of resource paths
+     */
+    Stream<ResourcePath> walk(ResourcePath path);
+
```

Comment:
I wouldn't go with the `walk` name, it can be confusing when using in combination with `ResourceManager`. You can also return the `Resource` directly when walking or just add variants for `Resource`s and `ResourcePath`s. A depth could also be useful to limit the directory structure complexity:
`Stream<ResourcePath> listResourcePaths(ResourcePath resourcePath)`
`Stream<ResourcePath> walkResourcePaths(ResourcePath resourcePath, int maxDepth)`
`Stream<ResourcePath> walkResourcePaths(ResourcePath resourcePath)`
`Stream<Resource> walkResources(ResourcePath resourcePath, int maxDepth);`
`Stream<Resource> walkResources(ResourcePath resourcePath)`

Reply:
What would be the difference between `listResourcePaths` and `walkResourcePaths`?